client/ayon_houdini/plugins/publish/collect_usd_layers.py

```diff
@@ -38,12 +39,17 @@ def copy_instance_data(instance_src, instance_dest, attr):
     keys = attr.split(".")
     for i, key in enumerate(keys):
         if key not in src_data:
-            break
+            raise PublishError(
+                message= f"key '{key}' does not exist on the source instance."
+            )
 
         src_value = src_data[key]
         if i != len(key):
             dest_data = dest_data.setdefault(key, {})
-            assert isinstance(dest_data, dict), "Destination must be a dict"
+            if not isinstance(dest_data, dict):
+                raise PublishError(
+                    message="Destination must be a dict."
+                )
```

Comment:
This provides way too little detail to be any meaningful error to any artist or alike. What destination where, for which key - and what did it get instead? We're really (I feel) going the wrong way if this can't just be a `TypeError` but requires a `PublishError` since we suddenly are introducing a super generic error much less clear? (I mean, still better than the assert, don't get me wrong)

@illicit thoughts?

Reply:
Should every error (that can be raised along the publishing) be a `PublishError` ?
I've copula of scattered thoughts. 
- Any error that is not `PublishError` or `PublishValidationError`  will cause the publisher UI to show the message (It's not your fault etc).
- Some errors are originated from user interactions e.g. not selecting the correct node or using a wrong colorspace. I believe these definitely should be `PublishError` and `PublishValidationError` .  These errors can be fixed by the user.
- Other errors like the one you highlighted above, honestly, I don't know why/how it can happen and if the user has something to do with it. I assume it requires a developer to deal with them. but using `PublishError` can help developers know about the problem from a screenshot. (Note: I didn't use the title in the `PublishError` so the publish plugin's label will be used.)

Reply:
> Any error that is not PublishError or PublishValidationError will cause the publisher UI to show the message (It's not your fault etc).

Yup - this was a _design decision_ at some point. I'm still not 100% confident that it's the right one, but it did force us into writing better `PublishValidationError`. If now we're just going to make EVERYTHING `PublishError` without providing better messages I think we're in the same problematic realm and we could basically revert to allow any error be nicely reported - we'll just still need to make sure that the majority of errors we raise are very clear what do to about them.

> Some errors are originated from user interactions e.g. not selecting the correct node or using a wrong colorspace. I believe these definitely should be PublishError and PublishValidationError . These errors can be fixed by the user.

I can't think of any now - but that does sound right. However, the error must be very clear, describe what is wrong and how to go about fixing it, etc.

> Other errors like the one you highlighted above, honestly, I don't know why/how it can happen and if the user has something to do with it. I assume it requires a developer to deal with them. but using PublishError can help developers know about the problem from a screenshot. (Note: I didn't use the title in the PublishError so the publish plugin's label will be used.)

Some assertions in the codebase were just there - to assert for a _potential_ edge case where it might happen... not necessarily that they would happen. That's why some might not necessarily be bad to remain asserts (or other regular errors). They were just there because we just really wanted to make sure if the code continued that something behaved a certain way.